src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ApplicationModelFactory.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                throw new ArgumentNullException(nameof(applicationModelProviders));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _applicationModelProviders = applicationModelProviders.OrderBy(p => p.Order).ToArray();
+            _conventions = options.Value.Conventions;
+        }
+
+        public ApplicationModel CreateApplicationModel(IEnumerable<TypeInfo> controllerTypes)
+        {
+            if (controllerTypes == null)
+            {
+                throw new ArgumentNullException(nameof(controllerTypes));
+            }
+
+            var context = new ApplicationModelProviderContext(controllerTypes);
+
+            for (var i = 0; i < _applicationModelProviders.Length; i++)
+            {
+                _applicationModelProviders[i].OnProvidersExecuting(context);
+            }
+
+            for (var i = _applicationModelProviders.Length - 1; i >= 0; i--)
+            {
+                _applicationModelProviders[i].OnProvidersExecuted(context);
+            }
+
+            ApplicationModelConventions.ApplyConventions(context.Result, _conventions);
+
+            return context.Result;
+        }
+
+        public static List<TResult> Flatten<TResult>(
+            ApplicationModel application,
+            Func<ApplicationModel, ControllerModel, ActionModel, SelectorModel, TResult> flattener)
+        {
+            var results = new List<TResult>();
+            var errors = new Dictionary<MethodInfo, IList<string>>();
+
+            var actionsByMethod = new Dictionary<MethodInfo, List<(ActionModel, SelectorModel)>>();
+            var actionsByRouteName = new Dictionary<string, List<(ActionModel, SelectorModel)>>(StringComparer.OrdinalIgnoreCase);
+
+            var routeTemplateErrors = new List<string>();
+
+            foreach (var controller in application.Controllers)
+            {
+                foreach (var action in controller.Actions)
+                {
+                    foreach (var selector in ActionAttributeRouteModel.FlattenSelectors(action))
+                    {
+                        // PostProcess attribute routes so we can observe any errors.
+                        ReplaceAttributeRouteTokens(controller, action, selector, routeTemplateErrors);
+
+                        // Add to the data structures we use to find errors.
+                        AddActionToMethodInfoMap(actionsByMethod, action, selector);
+                        AddActionToRouteNameMap(actionsByRouteName, action, selector);
+
+                        var result = flattener(application, controller, action, selector);
+                        Debug.Assert(result != null);
+
+                        results.Add(result);
+                    }
+                }
+            }
+
+            var attributeRoutingConfigurationErrors = new Dictionary<MethodInfo, string>();
```

Comment:
All of this validation code moved from `ControllerActionDescriptorProvider` and now operates on models instead of action descriptors. There's no functional difference, but it allows us to reuse the logic in my next PR.